src/StructuredLogger/BinaryLogger/BuildEventArgsReader.cs

```diff
@@ -203,10 +224,91 @@ private static bool IsAuxiliaryRecord(BinaryLogRecordKind recordKind)
-            if (kind == BinaryLogRecordKind.ProjectImportArchive && fileFormatVersion == 16)
+            var canHaveCorruptedSize = kind == BinaryLogRecordKind.ProjectImportArchive && fileFormatVersion == 16;
+            Stream embeddedStream = SliceOfEmdeddedContent(canHaveCorruptedSize);
+
+            if (ArchiveFileEncountered != null)
+            {
+                // We could create ZipArchive over the target stream, and write to that directly,
+                //  however, binlog format needs to know stream size upfront - which is unknown,
+                //  so we would need to write the size after - and that would require the target stream to be seekable (which it's not)
+                ProjectImportsCollector? projectImportsCollector = null;
+
+                if (EmbeddedContentRead != null)
+                {
+                    projectImportsCollector =
+                        new ProjectImportsCollector(Path.GetRandomFileName(), false, runOnBackground: false);
+                }
+
+                // We are intentionally not grace handling corrupt embedded stream
+
+                using var zipArchive = new ZipArchive(embeddedStream, ZipArchiveMode.Read);
+
+                foreach (var entry in zipArchive.Entries/*.OrderBy(e => e.LastWriteTime)*/)
+                {
+                    var file = ArchiveFile.From(entry, adjustPath: false);
+                    ArchiveFileEventArgs archiveFileEventArgs = new(file);
+                    ArchiveFileEncountered(archiveFileEventArgs);
+
+                    if (projectImportsCollector != null)
+                    {
+                        var resultFile = archiveFileEventArgs.ArchiveFile;
+
+                        projectImportsCollector.AddFileFromMemory(
+                            resultFile.FullPath,
+                            resultFile.Text,
+                            makePathAbsolute: false,
+                            entryCreationStamp: entry.LastWriteTime);
+                    }
+                }
+
+                // Once embedded files are replayed one by one - we can send the resulting stream to subscriber
+                if (OnBlobRead != null || EmbeddedContentRead != null)
+                {
+                    projectImportsCollector!.ProcessResult(
+                        streamToEmbed => InvokeEmbeddedDataListeners(kind, streamToEmbed),
+                        error => throw new InvalidDataException(error));
+                    projectImportsCollector.DeleteArchive();
+                }
+            }
+            else if (OnBlobRead != null || EmbeddedContentRead != null)
+            {
+                InvokeEmbeddedDataListeners(kind, embeddedStream);
+            }
+            else
+            {
+                embeddedStream.SkipBytes();
+            }
+        }
+
+        private void InvokeEmbeddedDataListeners(BinaryLogRecordKind kind, Stream embeddedStream)
+        {
+            byte[] bytes = null;

```

Comment:
Let's see who currently needs the giant byte array, perhaps we can get rid of it. It's a huge source of memory usage. Ideally I think an IReadOnlyList<ArchiveFile> is all the consumers should need.

I've been planning to optimize how we store the byte[] array and process it, but I decided to wait for your PR to land first to avoid causing a messy merge.

Reply:
I'm resolving this comment now, as the current way how embedded files are read is unchanged by this PR.

Changing all consumers to accept `Stream` instead would be great. It can get bit tricky - as it needs to be consumed at the time of invoking the event and cannot be shared by multiple consumers - both of which seem to be broken by current consuming patterns (but I've looked just very quickly).

Alternatively some tricks might possibly be used to allow lazy delayed consuming - they would boil down to reading from disk (either remembering position and size and then decompress and skip to the embedded file on demand, or storig the embedded archive into separate temporary file) - this can cause slower load of embedded files (but faster initial load of just events).

In ideal world we'd change the binlog format - the compressed events and then the archive would be each compressed independently and placed in a single file with short 'table of content' on start. That way we can read just the events during the initial load (faster -> win) and then when files would be needed, we can quickly switch to stream position with the archive and read files on demand (memory efficient -> win)